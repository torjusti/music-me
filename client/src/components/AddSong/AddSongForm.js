import React, { Component } from 'react';
import { Form, Modal, Button, Header, Message } from 'semantic-ui-react';
import Rate from '../Details/Rate';

const initialState = {
  // The currently entered rating.
  rating: undefined,

  // The currently entered title.
  title: '',

  // The currently entered artist.
  artist: '',

  // The currently entered album.
  album: '',

  // The currently entered genre.
  genre: '',

  // The currently entered description.

  description: '',

  // Whether or not the submit button has been clicked once, so
  // that errors can be shown.
  showErrors: false,
};

class AddSongForm extends Component {
  constructor(props) {
    super(props);

    this.state = initialState;

    if (props.song) {
      this.state = {
        ...this.state,
        ...props.song,
      };
    }
  }

  clearForm = () => {
    if (this.props.song) {
      return this.setState({
        ...initialState,
        ...this.props.song,
      });
    }

    this.setState(initialState);
  };

  handleSubmit = () => {
    const { title, artist, album, genre, description } = this.state;

    // Check if all required fields are filled.
    if (title && artist && album && genre && description) {
      this.setState({
        showErrors: false,
      });

      this.props.onSubmit({
        // We need to give undefined instead of null to the
        // serializer in order to give correct data to the server.
        rating: this.state.rating || undefined,
        title: this.state.title,
        artist: this.state.artist,
        album: this.state.album,
        genre: this.state.genre,
        description: this.state.description,
        id: this.state.id,
      });

      this.clearForm();

      return;
    }

    // Errors are remainding, so show theml.
    this.setState({
      showErrors: true,
    });
  };

  handleCancel = () => {
    this.clearForm();
    this.props.onClose();
  };

  render() {
    return (
      <Modal open={this.props.open} onClose={this.handleCancel}>
        <Header content={this.props.song ? 'Edit song' : 'Add song'} />

        <Modal.Content>
          <Form error={this.state.showErrors}>
            <Message
              error
              header="Unable to add song"
              content="Please fill in all required fields before submitting the form again."
            />

            <Form.Field
              label="Rating"
              control={Rate}
              rating={this.state.rating}
              onClick={rating => this.setState({ rating })}
            />

            <Form.Input
              fluid
              required
              label="Title"
              placeholder="Title"
              value={this.state.title}
              onChange={e => this.setState({ title: e.target.value })}
              error={this.state.showErrors && !this.state.title}
            />

            <Form.Input
              fluid
              required
              label="Artist"
              placeholder="Artist"
              value={this.state.artist}
              onChange={e => this.setState({ artist: e.target.value })}
              error={this.state.showErrors && !this.state.artist}
            />

            <Form.Input
              fluid
              required
              label="Album"
              placeholder="Album"
              value={this.state.album}
              onChange={e => this.setState({ album: e.target.value })}
              error={this.state.showErrors && !this.state.album}
            />

            <Form.Input
              fluid
              required
              label="Genre"
              placeholder="Genre"
              value={this.state.genre}
              onChange={e => this.setState({ genre: e.target.value })}
              error={this.state.showErrors && !this.state.genre}
            />

            <Form.TextArea
              required
              label="Description"
              placeholder="Description"
              value={this.state.description}
              onChange={e => this.setState({ description: e.target.value })}
              error={this.state.showErrors && !this.state.description}
            />
          </Form>
        </Modal.Content>

        <Modal.Actions>
          <Button color="red" onClick={this.handleCancel} content="Cancel" />

          <Button
            color="blue"
            onClick={this.handleSubmit}
            content={this.props.song ? 'Edit song' : 'Add song'}
          />
        </Modal.Actions>
      </Modal>
    );
  }
}

export default AddSongForm;
