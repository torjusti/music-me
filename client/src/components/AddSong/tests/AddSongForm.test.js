import React from 'react';
import { shallow } from 'enzyme';
import AddSongForm from '../AddSongForm';

describe('AddSongForm', () => {
  const wrapper = shallow(<AddSongForm />);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
