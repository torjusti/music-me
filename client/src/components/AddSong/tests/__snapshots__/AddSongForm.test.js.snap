// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSongForm renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddSongForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "centered": true,
      "children": Array [
        <Header
          content="Add song"
        />,
        <ModalContent>
          <Form
            as="form"
            error={false}
          >
            <Message
              content="Please fill in all required fields before submitting the form again."
              error={true}
              header="Unable to add song"
            />
            <FormField
              control={[Function]}
              label="Rating"
              onClick={[Function]}
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Title"
              maxLength={100}
              onChange={[Function]}
              placeholder="Title"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Artist"
              maxLength={100}
              onChange={[Function]}
              placeholder="Artist"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Album"
              maxLength={100}
              onChange={[Function]}
              placeholder="Album"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Genre"
              maxLength={100}
              onChange={[Function]}
              placeholder="Genre"
              required={true}
              value=""
            />
            <FormTextArea
              as={[Function]}
              control={[Function]}
              error={false}
              label="Description"
              onChange={[Function]}
              placeholder="Description"
              required={true}
              value=""
            />
          </Form>
        </ModalContent>,
        <ModalActions>
          <Button
            as="button"
            color="red"
            content="Cancel"
            role="button"
          />
          <Button
            as="button"
            color="blue"
            content="Save"
            onClick={[Function]}
            role="button"
          />
        </ModalActions>,
      ],
      "closeOnDimmerClick": true,
      "closeOnDocumentClick": false,
      "dimmer": true,
      "eventPool": "Modal",
      "onClose": undefined,
      "open": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "content": "Add song",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Form
            as="form"
            error={false}
          >
            <Message
              content="Please fill in all required fields before submitting the form again."
              error={true}
              header="Unable to add song"
            />
            <FormField
              control={[Function]}
              label="Rating"
              onClick={[Function]}
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Title"
              maxLength={100}
              onChange={[Function]}
              placeholder="Title"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Artist"
              maxLength={100}
              onChange={[Function]}
              placeholder="Artist"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Album"
              maxLength={100}
              onChange={[Function]}
              placeholder="Album"
              required={true}
              value=""
            />
            <FormInput
              as={[Function]}
              control={[Function]}
              error={false}
              fluid={true}
              label="Genre"
              maxLength={100}
              onChange={[Function]}
              placeholder="Genre"
              required={true}
              value=""
            />
            <FormTextArea
              as={[Function]}
              control={[Function]}
              error={false}
              label="Description"
              onChange={[Function]}
              placeholder="Description"
              required={true}
              value=""
            />
          </Form>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "form",
            "children": Array [
              <Message
                content="Please fill in all required fields before submitting the form again."
                error={true}
                header="Unable to add song"
              />,
              <FormField
                control={[Function]}
                label="Rating"
                onClick={[Function]}
              />,
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Title"
                maxLength={100}
                onChange={[Function]}
                placeholder="Title"
                required={true}
                value=""
              />,
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Artist"
                maxLength={100}
                onChange={[Function]}
                placeholder="Artist"
                required={true}
                value=""
              />,
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Album"
                maxLength={100}
                onChange={[Function]}
                placeholder="Album"
                required={true}
                value=""
              />,
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Genre"
                maxLength={100}
                onChange={[Function]}
                placeholder="Genre"
                required={true}
                value=""
              />,
              <FormTextArea
                as={[Function]}
                control={[Function]}
                error={false}
                label="Description"
                onChange={[Function]}
                placeholder="Description"
                required={true}
                value=""
              />,
            ],
            "error": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "content": "Please fill in all required fields before submitting the form again.",
                "error": true,
                "header": "Unable to add song",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "control": [Function],
                "label": "Rating",
                "onClick": [Function],
                "rating": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": [Function],
                "control": [Function],
                "error": false,
                "fluid": true,
                "label": "Title",
                "maxLength": 100,
                "onChange": [Function],
                "placeholder": "Title",
                "required": true,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": [Function],
                "control": [Function],
                "error": false,
                "fluid": true,
                "label": "Artist",
                "maxLength": 100,
                "onChange": [Function],
                "placeholder": "Artist",
                "required": true,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": [Function],
                "control": [Function],
                "error": false,
                "fluid": true,
                "label": "Album",
                "maxLength": 100,
                "onChange": [Function],
                "placeholder": "Album",
                "required": true,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": [Function],
                "control": [Function],
                "error": false,
                "fluid": true,
                "label": "Genre",
                "maxLength": 100,
                "onChange": [Function],
                "placeholder": "Genre",
                "required": true,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": [Function],
                "control": [Function],
                "error": false,
                "label": "Description",
                "onChange": [Function],
                "placeholder": "Description",
                "required": true,
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Button
              as="button"
              color="red"
              content="Cancel"
              role="button"
            />,
            <Button
              as="button"
              color="blue"
              content="Save"
              onClick={[Function]}
              role="button"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "color": "red",
              "content": "Cancel",
              "onClick": undefined,
              "role": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "color": "blue",
              "content": "Save",
              "onClick": [Function],
              "role": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "centered": true,
        "children": Array [
          <Header
            content="Add song"
          />,
          <ModalContent>
            <Form
              as="form"
              error={false}
            >
              <Message
                content="Please fill in all required fields before submitting the form again."
                error={true}
                header="Unable to add song"
              />
              <FormField
                control={[Function]}
                label="Rating"
                onClick={[Function]}
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Title"
                maxLength={100}
                onChange={[Function]}
                placeholder="Title"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Artist"
                maxLength={100}
                onChange={[Function]}
                placeholder="Artist"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Album"
                maxLength={100}
                onChange={[Function]}
                placeholder="Album"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Genre"
                maxLength={100}
                onChange={[Function]}
                placeholder="Genre"
                required={true}
                value=""
              />
              <FormTextArea
                as={[Function]}
                control={[Function]}
                error={false}
                label="Description"
                onChange={[Function]}
                placeholder="Description"
                required={true}
                value=""
              />
            </Form>
          </ModalContent>,
          <ModalActions>
            <Button
              as="button"
              color="red"
              content="Cancel"
              role="button"
            />
            <Button
              as="button"
              color="blue"
              content="Save"
              onClick={[Function]}
              role="button"
            />
          </ModalActions>,
        ],
        "closeOnDimmerClick": true,
        "closeOnDocumentClick": false,
        "dimmer": true,
        "eventPool": "Modal",
        "onClose": undefined,
        "open": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "content": "Add song",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Form
              as="form"
              error={false}
            >
              <Message
                content="Please fill in all required fields before submitting the form again."
                error={true}
                header="Unable to add song"
              />
              <FormField
                control={[Function]}
                label="Rating"
                onClick={[Function]}
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Title"
                maxLength={100}
                onChange={[Function]}
                placeholder="Title"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Artist"
                maxLength={100}
                onChange={[Function]}
                placeholder="Artist"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Album"
                maxLength={100}
                onChange={[Function]}
                placeholder="Album"
                required={true}
                value=""
              />
              <FormInput
                as={[Function]}
                control={[Function]}
                error={false}
                fluid={true}
                label="Genre"
                maxLength={100}
                onChange={[Function]}
                placeholder="Genre"
                required={true}
                value=""
              />
              <FormTextArea
                as={[Function]}
                control={[Function]}
                error={false}
                label="Description"
                onChange={[Function]}
                placeholder="Description"
                required={true}
                value=""
              />
            </Form>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "form",
              "children": Array [
                <Message
                  content="Please fill in all required fields before submitting the form again."
                  error={true}
                  header="Unable to add song"
                />,
                <FormField
                  control={[Function]}
                  label="Rating"
                  onClick={[Function]}
                />,
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  error={false}
                  fluid={true}
                  label="Title"
                  maxLength={100}
                  onChange={[Function]}
                  placeholder="Title"
                  required={true}
                  value=""
                />,
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  error={false}
                  fluid={true}
                  label="Artist"
                  maxLength={100}
                  onChange={[Function]}
                  placeholder="Artist"
                  required={true}
                  value=""
                />,
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  error={false}
                  fluid={true}
                  label="Album"
                  maxLength={100}
                  onChange={[Function]}
                  placeholder="Album"
                  required={true}
                  value=""
                />,
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  error={false}
                  fluid={true}
                  label="Genre"
                  maxLength={100}
                  onChange={[Function]}
                  placeholder="Genre"
                  required={true}
                  value=""
                />,
                <FormTextArea
                  as={[Function]}
                  control={[Function]}
                  error={false}
                  label="Description"
                  onChange={[Function]}
                  placeholder="Description"
                  required={true}
                  value=""
                />,
              ],
              "error": false,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "content": "Please fill in all required fields before submitting the form again.",
                  "error": true,
                  "header": "Unable to add song",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "control": [Function],
                  "label": "Rating",
                  "onClick": [Function],
                  "rating": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "error": false,
                  "fluid": true,
                  "label": "Title",
                  "maxLength": 100,
                  "onChange": [Function],
                  "placeholder": "Title",
                  "required": true,
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "error": false,
                  "fluid": true,
                  "label": "Artist",
                  "maxLength": 100,
                  "onChange": [Function],
                  "placeholder": "Artist",
                  "required": true,
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "error": false,
                  "fluid": true,
                  "label": "Album",
                  "maxLength": 100,
                  "onChange": [Function],
                  "placeholder": "Album",
                  "required": true,
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "error": false,
                  "fluid": true,
                  "label": "Genre",
                  "maxLength": 100,
                  "onChange": [Function],
                  "placeholder": "Genre",
                  "required": true,
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "error": false,
                  "label": "Description",
                  "onChange": [Function],
                  "placeholder": "Description",
                  "required": true,
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Button
                as="button"
                color="red"
                content="Cancel"
                role="button"
              />,
              <Button
                as="button"
                color="blue"
                content="Save"
                onClick={[Function]}
                role="button"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "color": "red",
                "content": "Cancel",
                "onClick": undefined,
                "role": "button",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "color": "blue",
                "content": "Save",
                "onClick": [Function],
                "role": "button",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
